SOPT-iOS/Projects/Domain/Sources/UseCase/HomeUseCase.swift

```diff
@@ -53,4 +59,26 @@ extension DefaultHomeUseCase: HomeUseCase {
             }
             .store(in: cancelBag)
     }
+    
+    public func getAppServices() {
+        repository.getAppServices()
+            .withUnretained(self)
+            .sink { event in
+                print("GetAppServices State: \(event)")
+            } receiveValue: { owner, services in
+                owner.appServices.send(services)
+            }
+            .store(in: cancelBag)
+    }
+    
```

Comment:
`func getAppServices()`  와 var `appServices = PassthroughSubject()` 를 구분한 이유는 무엇인가요?
아래와 같이 함수가 `AnyPublisher` 반환하는 방식에 대해선 어떻게 생각하시는지 궁금해요

```suggestion
    
    public func getAppServices() ->  AnyPublisher<[HomeAppServicesModel], Never>{
        repository.getAppServices()
            .catch { 어쩌구 }
            .eraseToAnyPublisher()
    }
    
```

Reply:
우선적인 이유는 기존 컨벤션에서 `PassthroughSubject`를 사용하고 있어서, 컨벤션을 따라가고 있긴 했는데요.

기존 방식의 장점은 `PassthroughSubject`가 방출해주는 값을 그대로 받기만 하면 되기 때문에, VM에서 해당 `getAppService()`를 호출만 하면 별다른 처리 없이 값을 가져올 수 있습니다. **변경된 값을 받아 상태 업데이트에만 집중하기 때문에, 책임 분리가 명확**하다고 생각했어요.

```swift
/// e.g., VM에서의 활용 예시
input.viewDidLoad
            .withUnretained(self)
            .sink { owner, _ in
                owner.useCase.getAppService()
            }.store(in: cancelBag)
```

적어주신 대로 **AnyPublisher를 반환하게 되면, 호출하는 쪽인 VM에서 해당 퍼블리셔를 따로 관리**해주어야 합니다. AnyPublisher를 반환값으로 받게 되면 여러 연산자를 활용해서 데이터를 처리할 수 있다는 장점이 있지만, 이미 저희 코드에서 비즈니스 로직이 분리되어 처리되구 있는데 굳이 VM에서 이 과정이 필요한가? **VM에서는 방출된 데이터만 받으면 되지 않을까?** 하는 생각입니다. 

제가 놓쳤거나 잘못 이해하고 있는 부분이 있다면 알려주세요..! 😊

Reply:
> AnyPublisher를 반환하게 되면, 호출하는 쪽인 VM에서 해당 퍼블리셔를 따로 관리

가 어떤 뜻일까욥?

Reply:
`PassthroughSubject`는 단순 값을 전달받기 때문에 VM에서 방출값을 받기만 하면 되는데, `AnyPublisher`를 반환해버리면 VM에서 아래와 같이 **반환된 퍼블리셔를 또 구독해야 하지 않나요**? (관리..라는 단어 선택이 오해의 소지가 있는 것 같아요🥹)

```swift
input.viewDidLoad
    .withUnretained(self)
    .sink { owner, _ in
        owner.useCase.getAppServices()
            .sink { services in
                // 처리
            }
            .store(in: cancelBag) 
    }
    .store(in: cancelBag)
```

Reply:
아하 재현님께서는 구독 속에 구독 문제를 피하고 싶었던 거 였군요!
flatMap을 통해 해당 문제를 해결할 수 있을 것 같아요

```swift
// 방법 1
input.viewDidLoad
    .flatMap { useCase.getAppServices() }
    .withUnretained(self)
    .sink { owner, services in
              
    }
    .store(in: cancelBag)
    
// 방법 2
input.viewDidLoad
    .flatMap(useCase.getAppServices)
    .withUnretained(self)
    .sink { owner, services in
              
    }
    .store(in: cancelBag)
```

방법 1은 클로저 안에 퍼블리셔를 넣는 방식이며,
방법 2는 swift에서 함수는 1급 객체임을 활용해서 인자로 넣어주는 방식이에요

Reply:
움 그렇긴 한데요...! 이미 UseCase에서 퍼블리셔를 구독해 데이터 스트림을 관리하고 있는데, 이중으로 퍼블리셔를 생성하고 VM에서 이를 구독해야 하는 이유가 있나요? flatMap을 활용하면 제가 드린 예시보다 간단하긴 하지만... VM 단에서는 UseCase가 방출한 데이터를 받기만 하면 되지 않나? 라는 생각이 들어서요.

Reply:
음 제 생각들을 정확하게 전달하기 위해 살짝은 T모드로 답변드리는 점 미리 양해부탁드려용
- AS-IS: UseCase에 PassThroughSubject 
- TO-BE: UseCase의 함수 리턴값 AnyPublisher


>  이중으로 퍼블리셔를 생성하고 VM에서 이를 구독해야 하는 이유가 있나요?
저는 flatMap 방식이 이중으로 퍼블리셔를 생성하는 방식이라고 생각하지 않아요. 오히려 하나의 데이터스트림으로 잇는 것이라고 생각해요.  저희 프로젝트 UC-Repository, Repository-Service 모두 AnyPublisher로 데이터스트림으로 하나로 잇고 있는 것 처럼요.
오히려 AS-IS에서 왜 뷰모델과 유즈케이스 관계에서 이 데이터스트림이 끊어져야 하는지에 대한 고민이 필요하다고 느껴요. (에러가 발생할 때를 생각해볼 수 있을 것 같아요.)

> VM 단에서는 UseCase가 방출한 데이터를 받기만 하면 되지 않나? 
해당 PR에서 제안드린 IndexSet로직을 처리한다고 가정해볼게요. (상상 로직이니 정확하지 않을 수 있음!). 저는 IndexSet은 뷰로직이기에 ViewModel까지만 알아야한다고 생각해요. (UseCase는 IndexSet에 대한 정보를 몰라야 해!)  UseCase에서 데이터가 반환됐을 때 해당 데이터와 IndexSet을 조합해서 ViewModel의 Output으로 방출해야하는 상황에선 VM은 단순히 UseCase의 데이터를 받기만 하는 역할보단 더 큰 책임이 있을 것 같아요.

Reply:
**함수 관점**
저는 함수를 **특정 입력에 대한 출력**이라고 생각해요. AS-IS 방식에선 `getAppService()`에 대한 입력값이 출력으로 나오지 않고 있어요. getAppService(앱 서비스들을 가져오다) 라는 함수에 대한 기대값이 Void인 것이 부자연스럽다고 느끼는 것 같아요.

**순수함수 관점**
Swift는 함수형 프로그래밍을 지원해요. 함수형 프로그래밍에선 순수함수란 개념이 중요하다고 생각해요. 물론 현실적인 이유로 모든 함수가 순수할 순 없지만 최대한 순수함수틱하게 함수를 작성하고 싶은 소망이 있습니다.
getAppService()는 함수 내부에서 passthroughSubject라는 함수 외부의 값을 수정하고 있기에, 함수 관점에서 바라본다면 기대하지 않은 동작이라고 생각이 들기도 하구용. 

이는 테스트 코드를 짤 때에도 고민이 생기는데요. AS-IS 방식으로 UseCase를 단위테스트 한다고 가정해보면,
`func  getAppService()`라는 함수의 기댓값은 `var appService`에 대한 관찰로 타당성을 증명해야 할 거에요.
이때 `func  getAppService()`와 `var appService`의 관계가 Swift 문법이 아닌 네이밍에 의존하고 있는 점이 불확실하다고 느껴집니다.

Reply:
ㅋㅋㅋ 아 괜찮습니닿 ㅋㅋ ㅋ 흥미롭네요...🥹

***

> 오히려, passthrough는 값을 send도 할 수 있기에 오히려 관리 범주가 늘어날 수도 있다고 생각해요

`PassthroughSubject` 의 특성에 대해서는 동의를 하는데요, 지금 구조상에서 `send()`가 남용되어 관리가 잘못될 우려가 있을만한 예시가 있을까요? 계층 구조가 잘 짜여있으면, 이 부분이 크게 문제되지 않을 것 같아서요.

>  왜 뷰모델과 유즈케이스 관계에서 이 데이터 스트림이 끊어져야 하는지에 대한 고민이 필요하다고 느껴요.

'현재 상태에서 데이터 스트림이 끊어져 있어서 단점이다' 라고 하기 보다는, UseCase가 단순 방출하는 값을 VM이 받는 방식으로 **책임 분리가 이루어져 있는 것**이라고 생각해요. 말씀하신 에러 처리에 대한 내용도 궁금한데요. 현재처럼 UseCase에서 에러를 캐치해서 처리하는 로직이 더 구조적으로 맞지 않나요?.?

> 저는 IndexSet은 뷰로직이기에 ViewModel까지만 알아야한다고 생각해요.

이 부분은 석우 님이 코리 달아주신 대로 고쳐도, UseCase가 IndexSet을 알게 될 일은 없을 것 같아요. UseCase로부터 받은 데이터를 IndexSet 기준으로 분류해 UI를 업데이트 해주는 거라서요! VM은 해당 데이터를 받으면, 단순히 그 부분의 IndexSet 값을 VC에 전달해주고 있습니다.

> 저는 함수를 특정 입력에 대한 출력이라고 생각해요. AS-IS 방식에선 getAppService()에 대한 입력값이 출력으로 나오지 않고 있어요. getAppService(앱 서비스들을 가져오다) 라는 함수에 대한 기대값이 Void인 것이 부자연스럽다고 느끼는 것 같아요.

`getAppService()`가 Void를 반환하는 건, 이 함수의 목적 자체가 명령형이기 때문에 어색하지 않다고 생각해요. 순수 함수를 지향하시는 점에서는 공감을 하는데요, 순수 함수를 최대한 작성하기 위해서 구독 관계를 이어야 한다 라고 연결짓는 것은 조금 어색한 것 같습니다. (하지만 TO-BE의 장점...! 이라고 생각합니다)

> 이는 테스트 코드를 짤 때에도 고민이 생기는데요. AS-IS 방식으로 UseCase를 단위테스트 한다고 가정해보면,
func  getAppService()라는 함수의 기댓값은 var appService에 대한 관찰로 타당성을 증명해야 할 거에요.
이때 func  getAppService()와 var appService의 관계가 Swift 문법이 아닌 네이밍에 의존하고 있는 점이 불확실하다고 느껴집니다.

네이밍에 의존하고 있다기 보다는, 명령을 내리는 함수의 역할과 이에 대한 변수가 가지는 맥락이 맞으면 괜찮지 않을까요? 단위 테스트에서도 명령에 대한 값을 해당 `PassThroughSubject` 변수가 잘 방출하는지 확인하면 될 것 같다는 생각이 들어요. 

제가 잘 몰라서 하는 궁금증인데요... 혹시 TO-BE로 단위 테스트를 진행했을 때, UseCase의 AnyPublisher가 반환한 데이터랑 뷰모델에서 구독한 부분까지 양쪽 모두 확인해야 하지 않나요?

*** 

조큼 길어지는 것 같아서..! 저희 회의 때 살짝 다시 이어서 얘기해보는 건 어떠실지..~ 제안드려봅니다 홀홀홀 ㅎ.ㅎ

Reply:
두 분 밤새 열띤 토론을 하셨네요🔥
저도 견해를 살짝 얹어보겠습니다,, 

> 저는 flatMap 방식이 이중으로 퍼블리셔를 생성하는 방식이라고 생각하지 않아요. 오히려 하나의 데이터스트림으로 잇는 것이라고 생각해요. 저희 프로젝트 UC-Repository, Repository-Service 모두 AnyPublisher로 데이터스트림으로 하나로 잇고 있는 것 처럼요.

우선 이 말에 적극 동의합니다! 그래서 AS-IS를 **UseCase에서 한 번 데이터 스트림을 끊어가는 방식**, TO-BE를 **하나의 데이터 흐름으로 처리하는 방식이라 생각**하고 의견을 적어볼게요.

일반적인 경우 `UseCase`와 `ViewModel`의 역할을 분리하는 것이 유연성과 확장성에 좋겠지만 `getAppServices()`의 경우처럼 데이터 가공이 하나도 없는 경우에도 역할 분리를 나누어야할까? 라는 의문에 동의합니다.

하지만 저희가 아키텍처를 설계하는 근본적인 이유를 생각해보자면 책임을 분리해야하기 때문이라는 재현님의 의견에도 동의가 돼요🤔

ㅋㅋ다 맞는말 같음 !! 

그래서 아키텍처를 왜 도입하는지 다시 한 번 짚어보았습니다.
아키텍처 없이도 시스템은 잘 돌아가지만, 많은 프로젝트에서 아키텍처를 도입하는 이유는 더 부드러운 시스템을 만들기 위함이잖아요?! 변경에 용이하고, , 뭐 그런.. 그렇기 때문에 책임과 역할을 열심히 분리하는 것이기도 하구요! 그리고 책임과 역할 분리는 일종의 방화벽 역할을 해준다고 생각합니다. (밑에 Entity와 DTO 리뷰의 경우처럼 !)

다시 `getAppServices()`의 경우로 돌아와서 생각해보았을 때 저는 **TO-BE의 방식이 조금 더 적합하다고 생각**합니다.
앱 서비스가 변경된다면 [HomeAppServiceModel]의 원소 개수가 달라질 순 있지만 저희가 HomeAppServiceModel 자체를 가공할 일이 없다고 생각하기 때문이에요.

그래서 유연성과 확장성을 고려해야할까?, 같은 element를 보내주는데 스트림을 한 번 끊을 필요가 있을까? 와 같은 의문이 듭니다.

| 혹시 TO-BE로 단위 테스트를 진행했을 때, UseCase의 AnyPublisher가 반환한 데이터랑 뷰모델에서 구독한 부분까지 양쪽 모두 확인해야 하지 않나요?
추가로 테스트 코드를 작성하게 될 때도 UseCase에서 AnyPublisher로 반환한 퍼블리셔와 뷰모델에서 구독한 퍼블리셔 동일하기 때문에 한 쪽만 확인해도 되어 오히려 편리하지 않을까 추측해보아요.

Reply:
저도 조심스럽게 의견 남겨봅니다 ㅎㅎ..

재현님과 석우님 두분의 의견이 너무 다 타당한 말씀인 것 같아요!
**저는 TO-BE를 아키텍처의 관점이 아니라 Combine의 관점으로 바라보았는데요.**

**Combine은 Type Erasure라는 장점이 있는 것 같아요!**
Publisher나 Subscriber 타입을 숨기고, 인터페이스를 추상화하려는 경우에 주로 사용되며,
특정 구현에 의존하지 않도록 할 수 있습니다. 그래서 확장성과 유연성이 높아지고, 구현채의 변경이 용이해지죠!

만약 TO-BE의 방식을 사용하게 된다면, AnyPublisher를 사용하면서 Publisher의 타입을 숨길 수 있을 겁니다.
구현체의 세부 사항에 의존하지 않을 수 있고, 내부 구현을 변경할 때 용이할 것 같아요. Publisher 타입을 명시적으로 다루지 않으면서 사용할 수 있고, 오히려 Passthrough을 사용했을 때보다 확장성과 변경에 용이할 것 같습니다!

마지막으로 캡슐화를 더 잘 지킬 수 있을 것 같아요 ㅎㅎ.. Passthrough을 사용할 때 처럼 외부에서 상태를 직접 수정할 수 있는 방법이 없어서 캡슐화에 이점이 있을 것 같다는 생각이 드네요!

> ++ 재현님 리뷰 중에 VM에서 해당 getAppService()를 호출만 하면 별다른 처리 없이 값을 가져올 수 있습니다. 
ViewModel에서 UseCase의 프로퍼티를 구독하여 데이터를 바인딩하고 있는 것 같아요.
결국 flatMap을 사용했을 때와 비슷한 처리를 하고 있다고 생각이 들어서 의견 남깁니다!

> ++ 스트림의 관점에서 
Combine이나 RxSwift 반응형 프로그램의 장점은 스트림을 용이하게 관리할 수 있다는 것 같아요.
TO-BE 방식이 하나의 흐름으로 관리를 하면서 조금 더 스트림 관리에 용이할 것 같다는 생각이 드네요!

> ++ 마지막으로 이중 구독하고 있는게 아닐까라는 생각이 듭니다!
ViewModel viewDidLoadEvent에서 sink안에 usecase의 네트워크 코드를 호출하는데,
repository.네트워크 메소드 안에서 sink를 사용하고 있는 것 같아요.
어떻게 보면 sink 안에서 sink를 호출하면서 불필요한 리소스가 발생할 수도 있을 것 같다는 생각이 듭니다!

Reply:
@meltsplit @yungu0010 @seungchan2 세 분 다 꼼꼼하게 리뷰 달아주셔서 감사합니다 🥹 잘못 이해하고 있었던 부분들이 있었던 것 같아 새롭게 배워가용 !! 리팩 방향에 대해서 추후에 논의해보면 좋을 것 같습니다 🙇🙇‍♀️🙇‍♂️

Reply:
코멘트를 다 읽어보진 않았지만 지나가다 재밌어보여서 짧은 의견 남겨보자면,
``` swift
  public func getAppServices() ->  AnyPublisher<[HomeAppServicesModel], Never> {}
```
이 형태의 인터페이스가 더 적절할 것 같아요.
대신, 이렇게 하나의 스트림으로 쭉 연결했을 때는 에러 발생 시 스트림이 끊어지게 되었을 때 이후에 유저가 버튼 터치 등의 액션을 했을 때 이벤트가 불러와지지 않을 수 있으니 에러 처리를 잘 하거나 스트림을 다시 잘 이어주는 처리가 필요할 것 같아요.

Reply:
@lsj8706 !!!!!! 코멘트 남겨주셔서 감사합니다 🥹